doc/designs/quic-design/stream-receive-buffers.md

```diff
@@ -0,0 +1,123 @@
+Stream Receive Buffers
+======================
+
+This is a QUIC specific module that handles the received stream data
+until the application reads it with SSL_read() or any future stream read
+calls.
+
+Receive Buffers requirements for MVP
+------------------------------------
+
+These are the requirements that were identified for MVP:
+
+- As packets with stream frames are received in arbitrary frames the
+  received data must be stored until all the data with earlier offsets
+  are received.
+- As packets can be received before application calls SSL_read() to read
+  the data the data must be stored.
+- The application should be able to set the limit on how much data should
+  be stored. The flow control should be used to limit the peer to not send
+  more data. Without the flow control limit a rogue peer could trigger
+  a DoS via unlimited flow of incoming stream data frames.
+- After the data is passed via SSL_read() to the application the stored
+  data can be released and flow control limit can be raised.
+- As the peer can recreate stream data frames when resending them, the
+  implementation must be able to handle properly frames with partially
+  or fully overlapping data with previously received frames.
+
+Optional Receive Buffers requirements
+-------------------------------------
+
+These are optional features of the stream receive buffers implementation.
+They are not required for MVP but they are otherwise desirable:
+
+- To support a single copy operation with a future stream read calls
+  the received data should not be copied out of the decrypted packets to
+  store the data. The only information actually stored would be a list
+  of offset, length, and pointers to data, along with a pointer to the
+  decrypted QUIC packet that stores the actual frame.
+
+Proposed new public API calls
+-----------------------------
+
+```C
+int SSL_set_max_stored_quic_stream_data(SSL *quic_stream, size_t length);
+```
+
+This function adjusts the current `quic_stream` data flow control limit to
+allow storing `length` of stream data before it is read by the application.
+OpenSSL handles sending MAX_STREAM_DATA frames appropriately when the
+application reads the stored data.
+
+```C
+int SSL_set_max_unprocessed_quic_packet_data(SSL *quic_connection,
+                                             size_t length);
```

Comment:
Same comment about keeping QUIC out of the API definition. Again this could plausibly impact DTLS. There is a buffer of unprocessed data there too, i.e. data intended for the next epoch (in master this buffer is implemented as a mem BIO).